test/slam_manifold_preintegration_update_test.cc

```diff
@@ -0,0 +1,314 @@
+
+  /**
+   * Returns a matrix whose (count) colums are samples from the distribution.
+   */
+  template <typename Generator>
+  Eigen::MatrixXd Sample(Generator& gen, const int count) {
+    Eigen::MatrixXd rand =
+        Eigen::MatrixXd::NullaryExpr(L.cols(), count, [&]() { return dist(gen); });
+    return L * rand;
+  }
+};
+
+/**
+ * Calculates the covariance of the columns of samples.
+ */
+Eigen::MatrixXd Covariance(const Eigen::MatrixXd& samples) {
+  const long int sample_size = samples.cols();
+  const Eigen::VectorXd avg_col = samples.rowwise().sum() / sample_size;
+  const Eigen::MatrixXd meanless_samples = samples.colwise() - avg_col;
+  return (meanless_samples * meanless_samples.transpose()) / (sample_size - 1);
+}
+
+TEST_CASE("Test MultiVarNormalDist generates samples with correct covariance", "[slam]") {
+  std::mt19937 gen(170);
+
+  // Create covar, a symmetric positive definite matrix
+  // NOTE(brad): This matrix is positive semi-definite because the diagonal entries all
+  // have values of at least 2.0, and the max value of an off-diagonal entry is .2. Thus
+  // within a row, the sum of the non-diagonal entries is capped by 0.2 * 8 = 1.6. Thus by the
+  // Gershgorin circle theorem, any eigen value of covar is at least 0.4 (i.e., positive).
+  std::uniform_real_distribution<double> uniform_dist(0.01, 0.1);
+  Eigen::MatrixXd covar = Eigen::MatrixXd::NullaryExpr(9, 9, [&]() { return uniform_dist(gen); });
+  covar += covar.transpose().eval();
+  covar.diagonal().array() += 2.0;
+
+  MultiVarNormalDist dist(covar);
+
+  const Eigen::MatrixXd samples = dist.Sample(gen, 1 << 23);
+  const Eigen::Matrix<double, 9, 9> calculated_covar = Covariance(samples);
+
+  for (int col = 0; col < 9; col++) {
+    for (int row = 0; row < 9; row++) {
+      const double covar_rc = covar(row, col);
+      const double difference = covar_rc - calculated_covar(row, col);
+      // NOTE(brad): trade-off between sample count (time) and accuracy of calculated_covar
+      CHECK(difference * difference < 9e-3 * covar_rc * covar_rc);
+    }
+  }
+}
+
+namespace example {
+static const Eigen::Vector3d accel_bias = {3.4, 1.6, -5.9};
+static const Eigen::Vector3d gyro_bias = {1.2, -2.4, 0.5};
+static const Eigen::Vector3d accel_cov(7e-5, 7e-5, 7e-5);
+static const Eigen::Vector3d gyro_cov(1e-3, 1e-3, 1e-3);
+static const Eigen::Vector3d true_gyro = Eigen::Vector3d::Constant(10.2);
+static const Eigen::Vector3d true_accel = Eigen::Vector3d::Constant(4.3);
+}  // namespace example
+
+TEST_CASE("Test ImuPreintegator.covariance", "[slam]") {
+  // In order to test that the computed covariance is correct, we sample the values many
+  // time to numerically calculate the covariance to compare.
+  using M99 = Eigen::Matrix<double, 9, 9>;
+  // Parameters
+  const double dt = 1e-3;
+  const int measurements_per_sample = 50;
+  const int sample_count = 1 << 18;
+
+  // Multithreading
+  const int thread_count = 16;
```

Comment:
This should not be 16 by default, it should really be 1

Reply:
I'll do this for now, but is there really that much of an overhead of spawning threads when only a single core is available? Especially if we're only spawning a fixed number and the process is kind of long.

Reply:
For this test probably not really, since CPU is the only resource that scales with number of threads here, and not memory.  In general though the test runner (ctest, bazel, etc) should have control over test parallelism and will run tests with whatever parallelism is requested there, and individual tests trying to parallelize themselves makes that less effective